ReplicaExchange/src/recore.jl

```diff
@@ -92,103 +99,194 @@ function re!(rng,
              samples,
              sim_chain_keep_n,
              sim_chain_keep_last = last ∘ sim_chain_keep_n,
-             swap_contexts! = swap_contexts!,
-             cb = tonothing)
+             swap_contexts! = swap_contexts!)
   nreplicas = length(states)
 
-
   GROUNDID = 1
   for num_swap = 1:num_swaps
-    for i = 1:nreplicas
-      @show i, :, logenergys[i](states[GROUNDID])
-      if i == GROUNDID
-        # @show typeof(logenergys)
-        # @assert false      
-        # If we're ground state, return swap_every samples
-        lb = (num_swap - 1) * samples_per_swap + 1
-        ub = lb + samples_per_swap - 1
-        @inbounds samples[lb:ub] = sim_chain_keep_n(rng,
-                                                    logenergys[GROUNDID],
-                                                    states[GROUNDID],
-                                                    samples_per_swap,
-                                                    i)
-        @inbounds states[GROUNDID] = samples[ub]
-      else
-        # If not ground state just return last sample
-        @inbounds states[i] = sim_chain_keep_last(rng, logenergys[i], states[i], samples_per_swap, i)
+    lb = (num_swap - 1) * samples_per_swap + 1
+    ub = lb + samples_per_swap - 1
+
+    Threads.@sync for i = 1:nreplicas
+      Threads.@spawn begin
+        if i == GROUNDID
+          # If we're ground state, return swap_every samples
+          @inbounds samples[lb:ub] = sim_chain_keep_n(rng,
+                                                  logenergys[GROUNDID],
+                                                  states[GROUNDID],
+                                                  samples_per_swap,
+                                                  i)
+          @inbounds states[GROUNDID] = samples[ub]
+        else
+          # If not ground state just return last sample
+          @inbounds states[i] = sim_chain_keep_last(rng, logenergys[i], states[i], samples_per_swap, i)
+        end
       end
     end
     swap_contexts!(rng, logenergys, states)
   end
   samples
 end
-# @pre length(logenergys) == length(states) "Must have one density per initial state"
 
-"Non-mutating re!"
+@pre re!(rng, logenergys, samples_per_swap, num_swaps, states, samples, sim_chain_keep_n) = num_swaps > 0 "There must be at least 1 swap"
+@pre re!(rng, logenergys, samples_per_swap, num_swaps, states, samples, sim_chain_keep_n) = all([isa(logenergy, Function) for logenergy in logenergys]) "logenergys is a Function"
```

Comment:
`isa(logenergy, Function)` does not need to be true (it's overly restrictive)

Reply:
a logenergy need only be some object on which logenergy(state) is defined, which may not be a subtype of `Function`